File: src/main/java/com/mercateo/jsonschema/property/PropertyBuilderDefault.java
package com.mercateo.jsonschema.property;

import com.mercateo.jsonschema.generictype.GenericType;
import com.mercateo.jsonschema.generictype.GenericTypeHierarchy;
import javaslang.collection.HashMap;
import javaslang.collection.HashSet;
import javaslang.collection.List;
import javaslang.collection.Map;
import javaslang.collection.Set;

import java.lang.annotation.Annotation;
import java.util.ArrayList;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

public final class PropertyBuilderDefault implements PropertyBuilder {
    private static final String ROOT_NAME = "#";

    private final GenericTypeHierarchy genericTypeHierarchy;

    private final List<RawPropertyCollector> rawPropertyCollectors;

    private final AnnotationMapBuilder annotationMapBuilder;

    private ConcurrentHashMap<GenericType<?>, PropertyDescriptor> knownDescriptors;

    public PropertyBuilderDefault(List<RawPropertyCollector> rawPropertyCollectors) {
        this.rawPropertyCollectors = rawPropertyCollectors;
        this.genericTypeHierarchy = new GenericTypeHierarchy();
        this.annotationMapBuilder = new AnnotationMapBuilder();
        this.knownDescriptors = new ConcurrentHashMap<>();
    }

    private static Object rootValueAccessor(Object object) {
        throw new IllegalStateException("cannot call value accessor for root element");
    }

    @Override
    public Property from(Class<?> propertyClass) {
        return from(GenericType.of(propertyClass));
    }

    @Override
    public Property from(GenericType<?> genericType) {
        return from(ROOT_NAME, genericType, HashMap.empty(),
                PropertyBuilderDefault::rootValueAccessor, HashSet.empty());
    }

    private Property from(String name, GenericType<?> genericType,
                          Map<Class<? extends Annotation>, Set<Annotation>> annotations, Function valueAccessor, Set<GenericType<?>> nestedTypes) {
        java.util.Map<GenericType<?>, PropertyDescriptor> addedDescriptors = new java.util.HashMap<>();
        java.util.Map<GenericType<?>, java.util.List<PropertyDescriptorReference>> addedReferences = new java.util.HashMap<>();
        final Property property = from(name, genericType, annotations, valueAccessor, addedDescriptors, addedReferences, nestedTypes);
        knownDescriptors.putAll(addedDescriptors);

        for (java.util.Map.Entry<GenericType<?>, java.util.List<PropertyDescriptorReference>> entry : addedReferences.entrySet()) {
            final PropertyDescriptor propertyDescriptor = knownDescriptors.get(entry.getKey());
            entry.getValue().forEach(propertyDescriptorReference -> {
                propertyDescriptorReference.setChildren(propertyDescriptor.children());
                propertyDescriptorReference.setReference("todo");
            });
        }
        return property;
    }

    private Property from(String name, GenericType<?> genericType, Map<Class<? extends Annotation>, Set<Annotation>> annotations, Function valueAccessor,
                          java.util.Map<GenericType<?>, PropertyDescriptor> addedDescriptors, java.util.Map<GenericType<?>, java.util.List<PropertyDescriptorReference>> addedReferences, Set<GenericType<?>> nestedTypes) {
        final PropertyDescriptor propertyDescriptor = getPropertyDescriptor(genericType, addedDescriptors, addedReferences, nestedTypes);

        return ImmutableProperty.of(name, propertyDescriptor, valueAccessor, annotationMapBuilder
                .merge(annotations, propertyDescriptor.annotations()));
    }

    private PropertyDescriptor getPropertyDescriptor(GenericType<?> genericType, java.util.Map<GenericType<?>, PropertyDescriptor> addedDescriptors, java.util.Map<GenericType<?>, java.util.List<PropertyDescriptorReference>> addedReferences, Set<GenericType<?>> nestedTypes) {
        if (knownDescriptors.containsKey(genericType)) {
            return knownDescriptors.get(genericType);
        } else {
            if (addedDescriptors.containsKey(genericType)) {
                return addedDescriptors.get(genericType);
            } else {
                if (!nestedTypes.contains(genericType)) {
                    return createPropertyDescriptor(genericType, addedDescriptors, addedReferences, nestedTypes);
                } else {
                    final PropertyType propertyType = PropertyTypeMapper.of(genericType);
                    final PropertyDescriptorReference propertyDescriptorReference = ImmutablePropertyDescriptorReference.of(propertyType, genericType,
                            annotationMapBuilder.createMap(genericType.getRawType().getAnnotations()));
                    addedReferences.computeIfAbsent(genericType, ignore -> new ArrayList()).add(propertyDescriptorReference);
                    return propertyDescriptorReference;
                }
            }
        }
    }

    private PropertyDescriptor createPropertyDescriptor(GenericType<?> genericType, java.util.Map<GenericType<?>, PropertyDescriptor> addedDescriptors, java.util.Map<GenericType<?>, java.util.List<PropertyDescriptorReference>> addedReferences, Set<GenericType<?>> nestedTypes) {
        final PropertyType propertyType = PropertyTypeMapper.of(genericType);

        final List<Property> children;
        switch (propertyType) {
            case OBJECT:
                children = createChildProperties(genericType, addedDescriptors, addedReferences, nestedTypes.add(genericType));
                break;

            case ARRAY:
                final Property property = from("", genericType.getContainedType(), HashMap.empty(), o -> null, addedDescriptors, addedReferences, nestedTypes);
                children = List.of(property);
                break;

            default:
                children = List.empty();
                break;
        }

        final PropertyDescriptor propertyDescriptor = ImmutablePropertyDescriptorDefault.of(propertyType, genericType, children, annotationMapBuilder.createMap(
                genericType.getRawType().getAnnotations()));
        addedDescriptors.put(genericType, propertyDescriptor);
        return propertyDescriptor;
    }

    private List<Property> createChildProperties(GenericType<?> genericType, java.util.Map<GenericType<?>, PropertyDescriptor> addedDescriptors, java.util.Map<GenericType<?>, java.util.List<PropertyDescriptorReference>> addedReferences, Set<GenericType<?>> nestedTypes) {
        return rawPropertyCollectors.flatMap(collector -> genericTypeHierarchy.hierarchy(
                genericType).flatMap(collector::forType)).map(rawProperty -> mapProperty(rawProperty, addedDescriptors, addedReferences, nestedTypes));
    }

    private Property mapProperty(RawProperty rawProperty, java.util.Map<GenericType<?>, PropertyDescriptor> addedDescriptors, java.util.Map<GenericType<?>, java.util.List<PropertyDescriptorReference>> addedReferences, Set<GenericType<?>> nestedTypes) {
        return from(rawProperty.name(), rawProperty.genericType(), rawProperty.annotations(),
                rawProperty.valueAccessor(), addedDescriptors, addedReferences, nestedTypes);
    }


}

File: src/main/java/com/mercateo/jsonschema/property/PropertyBuilderDefault.kt
package com.mercateo.jsonschema.property

import com.mercateo.jsonschema.collections.MutablePropertyDescriptorMap
import com.mercateo.jsonschema.generictype.GenericType
import com.mercateo.jsonschema.generictype.GenericTypeHierarchy
import java.util.*

class PropertyBuilderDefault(vararg rawPropertyCollectors: RawPropertyCollector) : PropertyBuilder {

    private val optionUnwrapper: Map<Class<*>, (Any) -> Any?> = mapOf(
            Pair(Optional::class.java, { option ->
                if (option is Optional<*>) {
                    option.orElse(null)
                } else {
                    null
                }
            })
    )

    private val rawPropertyCollectors: Array<out RawPropertyCollector> = rawPropertyCollectors

    private val genericTypeHierarchy: GenericTypeHierarchy = GenericTypeHierarchy()

    private val annotationMapBuilder: AnnotationMapBuilder = AnnotationMapBuilder()

    private val knownDescriptors: MutablePropertyDescriptorMap = MutablePropertyDescriptorMap()

    override fun <T> from(propertyClass: Class<T>): Property<Void, T> {
        return from(GenericType.of(propertyClass))
    }

    override fun <T> from(genericType: GenericType<T>): Property<Void, T> {
        return from(ROOT_NAME, genericType, mutableMapOf(),
                { it: Void -> rootValueAccessor(it) }, mutableSetOf())
    }

    private fun <S, T> from(
            name: String,
            genericType: GenericType<T>,
            annotations: Map<Class<out Annotation>, Set<Annotation>>,
            valueAccessor: (S) -> T?,
            nestedTypes: Set<GenericType<*>>): Property<S, T> {
        val addedDescriptors = MutablePropertyDescriptorMap()
        val property = from(name, genericType, annotations, valueAccessor, addedDescriptors, nestedTypes)
        knownDescriptors.putAll(addedDescriptors)

        return property
    }

    private fun <S, T> from(
            name: String,
            genericType: GenericType<T>,
            annotations: Map<Class<out Annotation>, Set<Annotation>>,
            valueAccessor: (S) -> T?,
            addedDescriptors: MutablePropertyDescriptorMap,
            nestedTypes: Set<GenericType<*>>): Property<S, T> {
        val propertyDescriptor = getPropertyDescriptor(genericType, addedDescriptors, nestedTypes)

        return Property(name, propertyDescriptor, valueAccessor, annotationMapBuilder.merge(annotations, propertyDescriptor.annotations))
    }

    private fun <T> getPropertyDescriptor(
            genericType: GenericType<T>,
            addedDescriptors: MutablePropertyDescriptorMap,
            nestedTypes: Set<GenericType<*>>): PropertyDescriptor<T> {
        if (knownDescriptors.containsKey(genericType)) {
            return knownDescriptors[genericType]
        } else {
            if (addedDescriptors.containsKey(genericType)) {
                return addedDescriptors[genericType]
            } else {
                if (!nestedTypes.contains(genericType)) {
                    return createPropertyDescriptor(genericType, addedDescriptors, nestedTypes)
                } else {
                    val propertyType = PropertyTypeMapper.of(genericType)
                    val annotations = genericType.rawType.annotations
                    return PropertyDescriptorDefault(propertyType, genericType, PropertyDescriptor.Context.InnerReference,
                            annotationMapBuilder.createMap(*annotations))
                }
            }
        }
    }

    private fun <T> createPropertyDescriptor(
            genericType: GenericType<T>,
            addedDescriptors: MutablePropertyDescriptorMap,
            nestedTypes: Set<GenericType<*>>
    ): PropertyDescriptor<T> {
        val propertyType = PropertyTypeMapper.of(genericType)

        val children: List<Property<T, Any>> = when (propertyType) {
            PropertyType.OBJECT -> createChildProperties(genericType, addedDescriptors, nestedTypes + genericType)

            PropertyType.ARRAY ->
                listOf(from("", genericType.containedType as GenericType<Any>, mutableMapOf(), { o -> null }, addedDescriptors, nestedTypes))

            else -> emptyList()
        }

        val annotations = genericType.rawType.annotations.filter { !it.annotationClass.qualifiedName!!.startsWith("kotlin.") }.toTypedArray()
        val propertyDescriptor = PropertyDescriptorDefault(propertyType, genericType, PropertyDescriptor.Context.Children(children), annotationMapBuilder.createMap(*annotations))
        addedDescriptors.put(genericType, propertyDescriptor)
        return propertyDescriptor
    }

    private fun <T> createChildProperties(
            genericType: GenericType<T>,
            addedDescriptors: MutablePropertyDescriptorMap,
            nestedTypes: Set<GenericType<*>>
    ): List<Property<T, Any>> {
        return rawPropertyCollectors.flatMap { collector ->
            genericTypeHierarchy.hierarchy(genericType).flatMap { collector.forType(it) }.map {
                it as RawProperty<T, Any>
            }.asIterable()
        }.map { mapProperty(it, addedDescriptors, nestedTypes) }
    }

    private fun <S, T> mapProperty(
            rawProperty: RawProperty<S, T>,
            addedDescriptors: MutablePropertyDescriptorMap,
            nestedTypes: Set<GenericType<*>>
    ): Property<S, Any> {
        val genericType = rawProperty.genericType
        val customUnwrapper = optionUnwrapper.get(genericType.rawType)

        val targetGenericType = (if (customUnwrapper == null) genericType else {
            genericType.containedType
        }) as GenericType<Any>

        val valueAccessor: (S) -> Any? = if (customUnwrapper == null) rawProperty.valueAccessor else { it: S ->
            val inner = rawProperty.valueAccessor.invoke(it)
            if (inner != null) customUnwrapper.invoke(inner) else null
        }
        return from(rawProperty.name,
                targetGenericType,
                rawProperty.annotations,
                valueAccessor, addedDescriptors, nestedTypes)
    }

    companion object {
        private val ROOT_NAME = "#"

        private fun <T> rootValueAccessor(instance: Void): T? {
            throw IllegalStateException("cannot call value accessor for root element")
        }
    }
}
